### PR TITLE
ddl: support create index with ingest method

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/35983

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/index.go [hunk 1]
```diff
@@ -675,20 +677,52 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
...
 	}
 	job.ReorgMeta.ReorgTp = model.ReorgTypeTxn
 	return model.ReorgTypeTxn
 }
 
+// Check if PiTR is enabled in cluster.
+func isPiTREnable(w *worker) bool {
+	ctx, err := w.sessPool.get()
+	if err != nil {
+		return true
```
Comment (wjhuang2016) [inline ddl/index.go:707]: Why return true directly?

File: ddl/ddl.go [hunk 2]
```diff
@@ -681,6 +682,8 @@ func (d *ddl) Start(ctxPool *pools.ResourcePool) error {
 	// Start some background routine to manage TiFlash replica.
 	d.wg.Run(d.PollTiFlashRoutine)
 
+	ingest.InitGlobalLightningEnv()
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/ddl.go:685]: Yes.

File: ddl/index.go (hunk 1 above)
Comment (tangenta) [reply ddl/index.go]: I've changed it to `canUseIngest()`

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/index.go [hunk 3]
```diff
@@ -704,25 +739,67 @@ func doReorgWorkForCreateIndexMultiSchema(w *worker, d *ddlCtx, t *meta.Meta, jo
...
 	case model.BackfillStateRunning:
-		done, ver, err = runReorgJobAndHandleErr(w, d, t, job, tbl, indexInfo, false)
-		if err != nil {
-			return false, ver, errors.Trace(err)
-		}
-		if done {
-			indexInfo.BackfillState = model.BackfillStateReadyToMerge
+		switch bfProcess {
+		case model.ReorgTypeLitMerge:
+			logutil.BgLogger().Info("[ddl] index backfill state running")
```
Comment (zimulala) [inline ddl/index.go:750]: Do we need to add "LitMerge" info to the log?

File: ddl/index.go [hunk 4]
```diff
@@ -704,25 +739,67 @@ func doReorgWorkForCreateIndexMultiSchema(w *worker, d *ddlCtx, t *meta.Meta, jo
...
+				break
+			}
+			if !ok && job.SnapshotVer != 0 {
+				// The owner is crashed or changed, we need to restart the backfill.
+				job.SnapshotVer = 0
+				return false, ver, nil
+			}
+			bc, err = ingest.LitBackCtxMgr.Register(w.ctx, indexInfo.Unique, job.ID, job.ReorgMeta.SQLMode)
+			if err != nil {
+				tryFallbackToTxnMerge(job, err)
```
Comment (zimulala) [inline ddl/index.go:762]: This function log "txn-merge", but this function is "lit-merge". Do we need to update this log?

File: ddl/index.go [hunk 5]
```diff
@@ -704,25 +739,67 @@ func doReorgWorkForCreateIndexMultiSchema(w *worker, d *ddlCtx, t *meta.Meta, jo
...
+				return false, ver, errors.Trace(err)
+			}
+			if !done {
+				return false, ver, nil
+			}
+			err = bc.FinishImport(indexInfo.ID, indexInfo.Unique, tbl)
+			if err != nil {
+				if kv.ErrKeyExists.Equal(err) {
+					logutil.BgLogger().Warn("import index duplicate key, convert job to rollback", zap.String("job", job.String()), zap.Error(err))
```
Comment (zimulala) [inline ddl/index.go:777]:
```suggestion
					logutil.BgLogger().Warn("[ddl] import index duplicate key, convert job to rollback", zap.String("job", job.String()), zap.Error(err))
```

File: ddl/index.go [hunk 6]
```diff
@@ -704,25 +739,67 @@ func doReorgWorkForCreateIndexMultiSchema(w *worker, d *ddlCtx, t *meta.Meta, jo
...
+					ver, err = convertAddIdxJob2RollbackJob(d, t, job, tbl.Meta(), indexInfo, err)
+				} else {
+					logutil.BgLogger().Warn("lightning import error", zap.Error(err))
```
Comment (zimulala) [inline ddl/index.go:780]:
```suggestion
					logutil.BgLogger().Warn("[ddl] lightning import error", zap.Error(err))
```

File: ddl/index.go (hunk 4 above)
Comment (tangenta) [reply ddl/index.go:762]: If there is an error for 'lit-merge', we will fallback to 'txn-merge'.

Comment (zimulala) [reply ddl/index.go:763]: Got it.

File: ddl/reorg.go [hunk 7]
```diff
@@ -253,9 +254,11 @@ func (w *worker) runReorgJob(rh *reorgHandler, reorgInfo *reorgInfo, tblInfo *mo
 		}
 
 		updateBackfillProgress(w, reorgInfo, tblInfo, 0)
-		if err1 := rh.RemoveDDLReorgHandle(job, reorgInfo.elements); err1 != nil {
-			logutil.BgLogger().Warn("[ddl] run reorg job done, removeDDLReorgHandle failed", zap.Error(err1))
-			return errors.Trace(err1)
+		if job.ReorgMeta.ReorgTp != model.ReorgTypeLitMerge || reorgInfo.mergingTmpIdx {
```
Comment (zimulala) [inline ddl/reorg.go:257]: Why we need to add `job.ReorgMeta.ReorgTp != model.ReorgTypeLitMerge` here?

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**